File: android/src/main/kotlin/com/anandhakrishnan/codeblock/downloader_flutter/DownloaderFlutterPlugin.kt
package com.anandhakrishnan.codeblock.downloader_flutter

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** DownloaderFlutterPlugin */
class DownloaderFlutterPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var singleFileDownloadChannel: MethodChannel
    private lateinit var multipleFileDownloadChannel: MethodChannel
    private lateinit var downloadProgressEventChannel: EventChannel
    val downloadSingleFileMethodChannel = "download_single_file_method_channel"
    val downloadSingleFileMethod = "downloadSingleFileMethod"
    val downloadMultipleFileMethodChannel = "download_multiple_file_method_channel"
    val downloadMultipleFileMethod = "downloadMultipleFileMethod"
    val downloadProgressEvent = "download_progress_events"
    private lateinit var context: Context
    private var eventSink: EventChannel.EventSink? = null


    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        singleFileDownloadChannel = MethodChannel(flutterPluginBinding.binaryMessenger, downloadSingleFileMethodChannel)
        singleFileDownloadChannel.setMethodCallHandler(this)
        multipleFileDownloadChannel = MethodChannel(flutterPluginBinding.binaryMessenger, downloadMultipleFileMethodChannel)
        multipleFileDownloadChannel.setMethodCallHandler(this)
        downloadProgressEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, downloadProgressEvent)
        downloadProgressEventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            downloadSingleFileMethod -> {
                val url = call.argument<String>("url")
                val fileName = call.argument<String>("file_name")
                val showToast = call.argument<Boolean>("show_toast") ?: false

                if (url.isNullOrEmpty()) {
                    result.error("INVALID_ARGUMENTS", "Missing 'url'", null)
                    return
                }

                if (fileName.isNullOrEmpty()){
                    result.error("INVALID_ARGUMENTS", "Missing 'file name", null)
                    return
                }

                // Call our new downloader class
                DownloadManagerFlutter.downloadSingleFile(
                    context,
                    url = url,
                    fileName = fileName,
                    showToast = showToast,
                    completion = { completion ->
                        result.success(completion)
                    },
                    progressCallback = { progressMap ->
                        eventSink?.success(progressMap)
                    })

            }

            downloadMultipleFileMethod -> {

                val urls = call.argument<List<String>>("urls")
                val fileNames = call.argument<List<String>>("file_names")
                val showToast = call.argument<Boolean>("show_toast") ?: false

                if (urls.isNullOrEmpty() || fileNames.isNullOrEmpty() || urls.size != fileNames.size) {
                    result.error("INVALID_ARGUMENTS", "URLs or file names missing/mismatched", null)
                    return
                }

                // Call our new downloader class
                DownloadManagerFlutter.downloadMultipleFiles(
                    context,
                    urls = urls,
                    fileNames = fileNames,
                    showToast = showToast,
                    completion = { completion ->
                        result.success(completion)
                    },
                    progressCallback = { progressMap ->
                        eventSink?.success(progressMap)
                    }
                )
            }
            
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        singleFileDownloadChannel.setMethodCallHandler(null)
        multipleFileDownloadChannel.setMethodCallHandler(null)
        downloadProgressEventChannel.setStreamHandler(null)
    }
}

File: ios/Classes/DownloaderFlutterPlugin.swift
import Flutter
import UIKit

public class DownloaderFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    var eventSink: FlutterEventSink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        
        let instance = DownloaderFlutterPlugin()
        
        let downloadSingleFileMethodChannel = FlutterMethodChannel(name: "download_single_file_method_channel",binaryMessenger:registrar.messenger())
        let downloadMultipleFileMethodChannel = FlutterMethodChannel(name: "download_multiple_file_method_channel",binaryMessenger:registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "download_progress_events", binaryMessenger: registrar.messenger())
        
        registrar.addMethodCallDelegate(instance, channel: downloadSingleFileMethodChannel)
        registrar.addMethodCallDelegate(instance, channel: downloadMultipleFileMethodChannel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let downloadManager = DownloadManager()
        switch call.method {
        case "downloadSingleFileMethod":
            if let args = call.arguments as? [String: Any],
               let url = args["url"] as? String,
               let fileName = args["file_name"] as? String,
                let saveToPhotos = args["save_to_photo"] as? Bool {
                print("📦 URL: \(url)")
                print("📄 File name: \(fileName)")
                
                downloadManager.downloadSingleFile(
                    from: url,
                    fileName: fileName,
                    saveToPhotos: saveToPhotos,
                    completion: { message in
                        result("\(message)")
                    },
                    progressCallback: { progress in
                        DispatchQueue.main.async {
                            self.eventSink?(progress)
                        }
                    }
                    
                )
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS",
                                    message: "Missing url or file_name",
                                    details: nil))
            }
            
        case "downloadMultipleFileMethod":
            if let args = call.arguments as? [String: Any],
               let urls = args["urls"] as? [String],
               let fileNames = args["file_names"] as? [String],
               let saveToPhotos = args["save_to_photo"] as? Bool {
                print("📦 URL: \(urls)")
                print("📄 File name: \(fileNames)")
                
                downloadManager.downloadMultipleFiles(
                    from: urls,
                    fileNames: fileNames,
                    saveToPhotos: saveToPhotos,
                    completion: { message in
                        result("\(message)")
                    },
                    progressCallback: { progress in
                        DispatchQueue.main.async {
                            self.eventSink?(progress)
                        }
                    }
                    
                )
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS",
                                    message: "Missing urls or file_names",
                                    details: nil))
            }
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
